dsa-solutions/lc-solutions/0000-0099/0025-reverse-nodes-in-k-group.md

```diff
@@ -0,0 +1,152 @@
+
+
+## Problem Description
+
+Given the head of a linked list, reverse the nodes of the list k at a time, and return the modified list.
+
+k is a positive integer and is less than or equal to the length of the linked list. If the number of nodes is not a multiple of k then left-out nodes, in the end, should remain as it is.
+
+You may not alter the values in the list's nodes, only nodes themselves may be changed.
+
+### Examples
+
+#### Example 1
+
+- **Input:** $head = [1,2,3,4,5], k = 2$
+- **Output:** $[2,1,4,3,5]$
+
+
+#### Example 2
+
+- **Input:** $head = [1,2,3,4,5], k = 3$
+- **Output:** $[3,2,1,4,5]$
+
+
+
+### Constraints
+
+- The number of nodes in the list is $n$.
+- $1 <= k <= n <= 5000$
+- $0 <= Node.val <= 1000$
+
+
+
+### Intuition
+
+The code aims to reverse nodes in a linked list in groups of $k$. It first checks if the length of the list is less than $k$ ; if so, it returns the head as-is. If the head is null or the list has only one node, or $k$ is less than 2, it returns the head. The core logic involves reversing the first $k$ nodes using a loop and recursively calling the function on the remaining nodes. The reversed portion is then linked with the result of the recursive call, ensuring the entire list is processed in $k$-sized groups.
+
+
+### Approach
+
+1. **Calculate Length of the Linked List:**
+
+    - Implement a helper function Length to calculate and return the size of the linked list. This function iterates through the list, incrementing a counter until the end of the list is reached.
+
+2. **Initial Checks:**
+
+    - In the reverseKGroup function, first check if $k$ is greater than the length of the list using the Length function. If so, return the head as is since there are not enough nodes to form a group.
+    - Check if the head is null, the list has only one node, or if $k$ is less than 2. If any of these conditions are true, return the head as-is since no reversal is needed.
+
+3. **Reverse First k Nodes:**
+
+    - Initialize three pointers: prev as null, curr as head, and next as null.
+    - Use a loop to reverse the first $k$ nodes. Within the loop:
+        - Store the next node of curr in next.
+        Reverse the link by setting curr->next to prev.
+        - Move the prev pointer to curr and curr to next.
+        - Increment the count to ensure only $k$ nodes are processed.
+
+3. **Recursive Call:**
+
+    - After reversing the first $k$ nodes, if next is not null (indicating there are more nodes left to process), make a recursive call to reverseKGroup with curr (the node following the first $k$ nodes) and $k$.
+    - Link the last node of the reversed group (which is now the head) to the result of the recursive call.
+
+4. **Return New Head:**
+
+    - Return prev as the new head of the reversed portion of the list.
+
+### Solution Code
+

```

Comment:
Please add code or if you not want to added then removed lines of empty code

````md
#### Python

```py

```

#### Java

```java

```
````